{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import collections\n",
    "import numpy as py\n",
    "import pandas as pd\n",
    "import itertools\n",
    "\n",
    "from ic_functions_py3 import*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symb</th>\n",
       "      <th>CG</th>\n",
       "      <th>FBgn</th>\n",
       "      <th>stage</th>\n",
       "      <th>staining</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>a10</td>\n",
       "      <td>CG6642</td>\n",
       "      <td>FBgn0011293</td>\n",
       "      <td>1</td>\n",
       "      <td>no staining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>a10</td>\n",
       "      <td>CG6642</td>\n",
       "      <td>FBgn0011293</td>\n",
       "      <td>2</td>\n",
       "      <td>no staining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a10</td>\n",
       "      <td>CG6642</td>\n",
       "      <td>FBgn0011293</td>\n",
       "      <td>3</td>\n",
       "      <td>no staining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>a10</td>\n",
       "      <td>CG6642</td>\n",
       "      <td>FBgn0011293</td>\n",
       "      <td>4</td>\n",
       "      <td>no staining</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>a10</td>\n",
       "      <td>CG6642</td>\n",
       "      <td>FBgn0011293</td>\n",
       "      <td>5</td>\n",
       "      <td>no staining</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_symb      CG         FBgn  stage     staining\n",
       "0       a10  CG6642  FBgn0011293      1  no staining\n",
       "1       a10  CG6642  FBgn0011293      2  no staining\n",
       "2       a10  CG6642  FBgn0011293      3  no staining\n",
       "3       a10  CG6642  FBgn0011293      4  no staining\n",
       "4       a10  CG6642  FBgn0011293      5  no staining"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create dictionaries of all associated names with the FlyBase gene symbol\n",
    "geneID,NametoIDs,NametoFBid = geneIDdictionary('Datasets/fbgn_annotation_ID_fb_2016_05.tsv')\n",
    "\n",
    "# Read in PFMs for promoter motifs and TFs\n",
    "pPFM,checkFBgn = readFASTA(\"Datasets/promoter_PFMs.fasta\",'PWM','v',geneID,0,0,0,0)\n",
    "PFM,checkFBgn = readFASTA(\"Datasets/Modified/PFMs.fasta\",'PWM','v',geneID,0,0,0,0)\n",
    "\n",
    "# Read in BDGP insitu data\n",
    "Stage = ['1_3','4_6','7_8','9_10','11_12','13_16']\n",
    "insitu = pd.read_csv('Datasets/insitu_annot.csv',names=['gene_symb','CG','FBgn','stage','staining'])\n",
    "insitu_ns = insitu[insitu.staining != 'no staining']\n",
    "insitu.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FBgn0283709 is in the cross references but not in geneIDs\n",
      "FBgn0267825 is in the cross references but not in geneIDs\n"
     ]
    }
   ],
   "source": [
    "# Read in all datasets\n",
    "# Cross-referencing dictionary for promoters\n",
    "crossref = pd.read_csv('Datasets/EPD/cross_references.txt',header=0,sep='\\t')\n",
    "\n",
    "# Promoters (-50 to +50)\n",
    "raw_promoters = readFASTA('Datasets/EPD/dm6promoters_minus50plus50.fa','sequences',0,geneID,re.compile('>FP\\d+\\s([\\w\\-\\(\\)\\.]+_\\d)'))\n",
    "promoters = {}\n",
    "title = []\n",
    "neither = []\n",
    "for k,v in raw_promoters.items():\n",
    "    number = k[-2:]\n",
    "    if k[:-2] in geneID.keys(): \n",
    "        promoters[geneID[k[:-2]][0]+number] = v\n",
    "    elif k[:-2].lower() in geneID.keys(): \n",
    "        promoters[geneID[k[:-2].lower()][0]+number] = v\n",
    "    elif k[:-2].title() in geneID.keys(): \n",
    "        promoters[geneID[k[:-2].title()][0]+number] = v\n",
    "        title.append(k[:-2].title())\n",
    "    elif not crossref['Ensembl Gene ID'][crossref['Associated Gene Name'] == k[:-2]].empty:\n",
    "        FBgn = crossref['Ensembl Gene ID'][crossref['Associated Gene Name'] == k[:-2]].values[0]\n",
    "        if FBgn in geneID.keys():\n",
    "            promoters[geneID[FBgn][0]+number] = v\n",
    "        else:\n",
    "            print('%s is in the cross references but not in geneIDs' %FBgn)\n",
    "    else:\n",
    "        neither.append(k)\n",
    "        # note: these genes seem to be withdrawn, in progress, or not found in FlyBase\n",
    "# Note: Both FBgn0283709 and FBgn0267825 are unannotated\n",
    "od_promoters = collections.OrderedDict(sorted(promoters.items()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>VTID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>FlyBase ID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Chromosome.1</th>\n",
       "      <th>Start.1</th>\n",
       "      <th>End.1</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VT0006</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>16836</td>\n",
       "      <td>18924</td>\n",
       "      <td>FBgn0002121</td>\n",
       "      <td>l(2)gl</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>9839</td>\n",
       "      <td>21376</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>VT0025</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>54047</td>\n",
       "      <td>56104</td>\n",
       "      <td>FBgn0051973</td>\n",
       "      <td>Cda5</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>25402</td>\n",
       "      <td>65404</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VT0131</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>285217</td>\n",
       "      <td>287380</td>\n",
       "      <td>FBgn0031245</td>\n",
       "      <td>CG3625</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>283385</td>\n",
       "      <td>291795</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>VT0132</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>288875</td>\n",
       "      <td>290893</td>\n",
       "      <td>FBgn0025686</td>\n",
       "      <td>Amnionless</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>287252</td>\n",
       "      <td>289144</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VT0132</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>288875</td>\n",
       "      <td>290893</td>\n",
       "      <td>FBgn0031245</td>\n",
       "      <td>CG3625</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>283385</td>\n",
       "      <td>291795</td>\n",
       "      <td>-</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     VTID Chromosome   Start     End   FlyBase ID      Symbol Chromosome.1  \\\n",
       "0  VT0006      chr2L   16836   18924  FBgn0002121      l(2)gl        chr2L   \n",
       "1  VT0025      chr2L   54047   56104  FBgn0051973        Cda5        chr2L   \n",
       "2  VT0131      chr2L  285217  287380  FBgn0031245      CG3625        chr2L   \n",
       "3  VT0132      chr2L  288875  290893  FBgn0025686  Amnionless        chr2L   \n",
       "4  VT0132      chr2L  288875  290893  FBgn0031245      CG3625        chr2L   \n",
       "\n",
       "   Start.1   End.1 Orientation  Match  \n",
       "0     9839   21376           -      2  \n",
       "1    25402   65404           -      2  \n",
       "2   283385  291795           -      2  \n",
       "3   287252  289144           -      2  \n",
       "4   283385  291795           -      1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in Vienna Tile enhancers\n",
    "VT = pd.read_csv('Datasets/Stark_ViennaTile/nature13395-s2/2014-01-00083C-Supplementary Table 4.csv')\n",
    "VT.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Source</th>\n",
       "      <th>Name</th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>assigned gene</th>\n",
       "      <th>FBGn</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>REDFly</td>\n",
       "      <td>ci_7.1</td>\n",
       "      <td>4</td>\n",
       "      <td>76479</td>\n",
       "      <td>83352</td>\n",
       "      <td>ci</td>\n",
       "      <td>FBgn0004859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAD</td>\n",
       "      <td>ey_UE2.0</td>\n",
       "      <td>4</td>\n",
       "      <td>720530</td>\n",
       "      <td>722315</td>\n",
       "      <td>ey</td>\n",
       "      <td>FBgn0005558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAD</td>\n",
       "      <td>ey_UE0.9</td>\n",
       "      <td>4</td>\n",
       "      <td>724596</td>\n",
       "      <td>725290</td>\n",
       "      <td>ey</td>\n",
       "      <td>FBgn0005558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAD</td>\n",
       "      <td>ey_DO2</td>\n",
       "      <td>4</td>\n",
       "      <td>730630</td>\n",
       "      <td>730841</td>\n",
       "      <td>ey</td>\n",
       "      <td>FBgn0005558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>REDFly</td>\n",
       "      <td>sphinx_1067bp_5'_fragment</td>\n",
       "      <td>4</td>\n",
       "      <td>994776</td>\n",
       "      <td>995842</td>\n",
       "      <td>sphinx</td>\n",
       "      <td>FBgn0083990</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Source                       Name Chr   Start    Stop assigned gene  \\\n",
       "0  REDFly                     ci_7.1   4   76479   83352            ci   \n",
       "1     CAD                   ey_UE2.0   4  720530  722315            ey   \n",
       "2     CAD                   ey_UE0.9   4  724596  725290            ey   \n",
       "3     CAD                     ey_DO2   4  730630  730841            ey   \n",
       "4  REDFly  sphinx_1067bp_5'_fragment   4  994776  995842        sphinx   \n",
       "\n",
       "          FBGn  \n",
       "0  FBgn0004859  \n",
       "1  FBgn0005558  \n",
       "2  FBgn0005558  \n",
       "3  FBgn0005558  \n",
       "4  FBgn0083990  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in CAD/RedFly \"known\" enhancers\n",
    "Known = pd.read_excel('Datasets/Furlong_4C/nature13417/nature13417-s3.xlsx')\n",
    "Known.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>view</th>\n",
       "      <th>cond</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>midpoint</th>\n",
       "      <th>mid_dist</th>\n",
       "      <th>max_dist</th>\n",
       "      <th>short_dist</th>\n",
       "      <th>zscore</th>\n",
       "      <th>feature_id</th>\n",
       "      <th>feature</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CRM_1088</td>\n",
       "      <td>MESO_3-4h</td>\n",
       "      <td>0.001</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>15164485</td>\n",
       "      <td>15165376</td>\n",
       "      <td>15164930</td>\n",
       "      <td>-135286</td>\n",
       "      <td>-135286</td>\n",
       "      <td>-134840</td>\n",
       "      <td>4.194707</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CRM_1088</td>\n",
       "      <td>MESO_3-4h</td>\n",
       "      <td>0.001</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>15182513</td>\n",
       "      <td>15183814</td>\n",
       "      <td>15183164</td>\n",
       "      <td>-117052</td>\n",
       "      <td>-117052</td>\n",
       "      <td>-116402</td>\n",
       "      <td>4.234489</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRM_1088</td>\n",
       "      <td>MESO_3-4h</td>\n",
       "      <td>0.001</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>15191016</td>\n",
       "      <td>15191709</td>\n",
       "      <td>15191362</td>\n",
       "      <td>-108854</td>\n",
       "      <td>-108854</td>\n",
       "      <td>-108507</td>\n",
       "      <td>3.489552</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CRM_1088</td>\n",
       "      <td>MESO_3-4h</td>\n",
       "      <td>0.001</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>15197589</td>\n",
       "      <td>15198942</td>\n",
       "      <td>15198266</td>\n",
       "      <td>-101950</td>\n",
       "      <td>-101950</td>\n",
       "      <td>-101274</td>\n",
       "      <td>3.378263</td>\n",
       "      <td>FBgn0034433</td>\n",
       "      <td>promoter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CRM_1088</td>\n",
       "      <td>MESO_3-4h</td>\n",
       "      <td>0.001</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>15206172</td>\n",
       "      <td>15206872</td>\n",
       "      <td>15206522</td>\n",
       "      <td>-93694</td>\n",
       "      <td>-93694</td>\n",
       "      <td>-93344</td>\n",
       "      <td>3.572291</td>\n",
       "      <td>NaN</td>\n",
       "      <td>intragenic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       view       cond  pvalue    chr     start       end  midpoint  mid_dist  \\\n",
       "0  CRM_1088  MESO_3-4h   0.001  chr2R  15164485  15165376  15164930   -135286   \n",
       "1  CRM_1088  MESO_3-4h   0.001  chr2R  15182513  15183814  15183164   -117052   \n",
       "2  CRM_1088  MESO_3-4h   0.001  chr2R  15191016  15191709  15191362   -108854   \n",
       "3  CRM_1088  MESO_3-4h   0.001  chr2R  15197589  15198942  15198266   -101950   \n",
       "4  CRM_1088  MESO_3-4h   0.001  chr2R  15206172  15206872  15206522    -93694   \n",
       "\n",
       "   max_dist  short_dist    zscore   feature_id     feature  \n",
       "0   -135286     -134840  4.194707          NaN  intragenic  \n",
       "1   -117052     -116402  4.234489          NaN  intragenic  \n",
       "2   -108854     -108507  3.489552          NaN  intragenic  \n",
       "3   -101950     -101274  3.378263  FBgn0034433    promoter  \n",
       "4    -93694      -93344  3.572291          NaN  intragenic  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in Furlong 4C enhancers\n",
    "Furlong4C = pd.read_excel('Datasets/Furlong_4C/nature13417/nature13417-s2.xlsx')\n",
    "Furlong4C.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>motif</th>\n",
       "      <th>position</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ECSIT_1</td>\n",
       "      <td>BREd</td>\n",
       "      <td>71</td>\n",
       "      <td>0.6065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fhos_4</td>\n",
       "      <td>BREd</td>\n",
       "      <td>32</td>\n",
       "      <td>0.9317</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CG33552_1</td>\n",
       "      <td>BREd</td>\n",
       "      <td>58</td>\n",
       "      <td>0.5349</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CG33552_1</td>\n",
       "      <td>BREd</td>\n",
       "      <td>61</td>\n",
       "      <td>0.7859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mrt_2</td>\n",
       "      <td>BREd</td>\n",
       "      <td>63</td>\n",
       "      <td>0.5518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sequence motif  position   score\n",
       "0    ECSIT_1  BREd        71  0.6065\n",
       "1     Fhos_4  BREd        32  0.9317\n",
       "2  CG33552_1  BREd        58  0.5349\n",
       "3  CG33552_1  BREd        61  0.7859\n",
       "4      mrt_2  BREd        63  0.5518"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find core promoter motifs\n",
    "_,_,filenames = next(os.walk('matrix'),(None, None, []))\n",
    "pmotif_list = [x[:-4] for x in filenames if '.txt' in x]\n",
    "\n",
    "ES_cutoff=dict(zip(pmotif_list,[0.5,0.05,0.01,0.01,0.01,0.01,0.1,0.01,0.01,0.01,0.01]))\n",
    "ppatser_output = ElemeNT_scoring(promoters,'promoters',pmotif_list,pPFM,ES_cutoff,os.getcwd())\n",
    "ppatser_output.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
